### PR TITLE
feat: test cases for retrive_tools functions

### PR DESCRIPTION
**Pull request by Hither1**
## Description

Retrieve tools function and unit test.

## Motivation and Context

Tools search https://github.com/camel-ai/camel/issues/569

## Types of changes

- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Wendong-Fan** [comment]
Hey @Hither1 , thanks for the contribution! Is this PR under developing or it's ready for review?

File: camel/functions/__init__.py [hunk 1]
```diff
@@ -26,6 +26,9 @@
 from .twitter_function import TWITTER_FUNCS
 from .weather_functions import WEATHER_FUNCS
 from .slack_functions import SLACK_FUNCS
+# from .retrieve_tools_function import RETRIEVE_TOOLS_FUNCS
```
Comment (Wendong-Fan) [inline camel/functions/__init__.py:29]: here need to be cleaned

File: camel/functions/__init__.py [hunk 2]
```diff
@@ -38,5 +41,6 @@
     'SEARCH_FUNCS',
     'TWITTER_FUNCS',
     'WEATHER_FUNCS',
+    # 'RETRIEVE_TOOLS_FUNCS',
```
Comment (Wendong-Fan) [inline camel/functions/__init__.py:44]: same as above

File: camel/functions/retrieve_tools_function.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+from camel.functions.openai_function import OpenAIFunction
+
+
+def embed_text(text: Optional[str], model, tokenizer):
+    inputs = tokenizer(text, return_tensors="pt")
+    with torch.no_grad():
+        outputs = model(**inputs)
+
+    return outputs.last_hidden_state.mean(dim=1).squeeze().cpu().numpy()
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:28]: 1. `text` is required
2. set type hint for the args
3. docstring is missing

File: camel/functions/retrieve_tools_function.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def cosine_similarity(vec1: np.ndarray, vec2: np.ndarray) -> float:
+    return np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:32]: what if one of the input vector's size is zero? add more error handling to make the code more robust

File: camel/functions/retrieve_tools_function.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+def retrieve_tools(
+    query: str, tools: List[OpenAIFunction], k: int = 3
+) -> Optional[List[OpenAIFunction]]:
+    r"""Using semantic search to retrieve relevent tools
+    OpenAI / Open source Embedding models to embed the query and the schemas / docstring of `OpenAIFunction`
+    Find the most relevent ones based on their embedding vectors
+
+    Args:
+        tools (List): The tools to be searched.
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:43]: `query` and `k` is missing

File: camel/functions/retrieve_tools_function.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#     k=2,
+# )
+# print(retrieved_tools)
+# get_weather_data, search_google_and_summarize
+
+
+# RETRIEVE_TOOLS_FUNCS: List[OpenAIFunction] = [
+#     OpenAIFunction(func)  # type: ignore[arg-type]
+#     for func in [retrieve_tools]
+# ]
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:81]: need to be cleaned

File: camel/functions/retrieve_tools_function.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    for func in tools:
+        func_embedding = embed_text(func.func.__doc__, model, tokenizer)
+        similarity = cosine_similarity(query_embedding, func_embedding)
+        similarities.append((similarity, func))
+
+    similarities.sort(key=lambda x: x[0], reverse=True)
+    if similarities[0][0] < 0.4:
+        return None
+    else:
+        retrieved_tools = [func for _, func in similarities[:k]]
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:63]: `k` value is not limited to be `int`, also limit it >0?

File: camel/functions/retrieve_tools_function.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    query_embedding = embed_text(query, model, tokenizer)
+
+    similarities = []
+    for func in tools:
+        func_embedding = embed_text(func.func.__doc__, model, tokenizer)
+        similarity = cosine_similarity(query_embedding, func_embedding)
+        similarities.append((similarity, func))
+
+    similarities.sort(key=lambda x: x[0], reverse=True)
+    if similarities[0][0] < 0.4:
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:60]: make the similarity score as a customizable arg?

File: camel/functions/retrieve_tools_function.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        retrieved_tools (List): List of retrieved (most relevant) tools.
+    """
+    tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+    model = AutoModel.from_pretrained("bert-base-uncased")
+
+    query_embedding = embed_text(query, model, tokenizer)
+
+    similarities = []
+    for func in tools:
+        func_embedding = embed_text(func.func.__doc__, model, tokenizer)
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:55]: add handling for cases when the tools list is empty or all provided descriptions are `None`?

File: camel/functions/retrieve_tools_function.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    Find the most relevent ones based on their embedding vectors
+
+    Args:
+        tools (List): The tools to be searched.
+
+    Returns:
+        retrieved_tools (List): List of retrieved (most relevant) tools.
+    """
+    tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+    model = AutoModel.from_pretrained("bert-base-uncased")
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:49]: move model and tokenizer initialization outside of retrieve_tools to improve efficiency?

File: test/functions/test_search_functions.py
Comment (Wendong-Fan) [inline test/functions/test_search_functions.py:1]: why modify test file for search functions here?

**Review by Wendong-Fan — CHANGES_REQUESTED** [review]
Thanks for the contribution! The functionality is good. But there is room for improvement in terms of code formatting, robustness, and documentation.

File: camel/functions/retrieve_tools_function.py (hunk 8 above)
Comment (Hither1) [reply camel/functions/retrieve_tools_function.py:60]: Hi @Wendong-Fan  I don't quite understand what you mean here

File: test/functions/test_search_functions.py
Comment (Hither1) [reply test/functions/test_search_functions.py:1]: Actually I didn't intend to, can you let me know which version of search function I should use to correct this?

File: camel/functions/retrieve_tools_function.py (hunk 10 above)
Comment (Hither1) [reply camel/functions/retrieve_tools_function.py:49]: Do you mean to put them as global variables of the file?

File: camel/functions/retrieve_tools_function.py (hunk 8 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:60]: here you hardcoded the similarity score to 0.4, it would be better to let user decide the threshold

File: camel/functions/retrieve_tools_function.py (hunk 10 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:49]: It's ok for current code structure, IMO the idea solution is to make  retrive_tools functions as one class and do the model and tokenizer initialization in __init__, we will do the refactor move all functions to toolkits, refer to https://github.com/camel-ai/camel/issues/610 , let's keep your current design here for now~

File: test/functions/test_search_functions.py
Comment (Wendong-Fan) [reply test/functions/test_search_functions.py:1]: Screenshot 2024-06-07 at 18 55 49

Your PR is modifying the test file for search functions, if it's a behaviour by mistake you can revert all the change there

File: camel/functions/__init__.py (hunk 1 above)
Comment (Wendong-Fan) [reply camel/functions/__init__.py:29]: hasn't cleaned it

File: camel/functions/__init__.py (hunk 2 above)
Comment (Wendong-Fan) [reply camel/functions/__init__.py:44]: same

File: camel/functions/retrieve_tools_function.py (hunk 3 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:28]: hasn't updated

File: camel/functions/retrieve_tools_function.py (hunk 4 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:32]: hasn't updated

File: camel/functions/retrieve_tools_function.py (hunk 5 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:43]: hasn't updated

File: camel/functions/retrieve_tools_function.py (hunk 9 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:55]: hasn't updated

File: camel/functions/retrieve_tools_function.py (hunk 6 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:81]: you can add example file like how https://github.com/camel-ai/camel/pull/612 did

File: camel/functions/retrieve_tools_function.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+import torch
+from transformers import AutoModel, AutoTokenizer
+
+from camel.functions.openai_function import OpenAIFunction
+
+
+def embed_text(text: str, model: AutoModel, tokenizer: AutoTokenizer):
+    r"""
+    Embed text using chosen model and tokenizer.
+    """
```
Comment (yiyiyi0817) [inline camel/functions/retrieve_tools_function.py:26]: Add more detailed docstring, such as Google style docstring as other functions in camel, in order to help others to maintain this function in the future?

File: camel/functions/retrieve_tools_function.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    inputs = tokenizer(text, return_tensors="pt")
+    with torch.no_grad():
+        outputs = model(**inputs)
+
+    return outputs.last_hidden_state.mean(dim=1).squeeze().cpu().numpy()
+
+
+def cosine_similarity(vec1: np.ndarray, vec2: np.ndarray) -> float:
```
Comment (yiyiyi0817) [inline camel/functions/retrieve_tools_function.py:34]: Same as above.

File: camel/functions/retrieve_tools_function.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        return np.dot(vec1, vec2) / (
+            np.linalg.norm(vec1) * np.linalg.norm(vec2)
+        )
+
+
+def retrieve_tools(
+    query: str, tools: List[OpenAIFunction], k: int = 3, sim_thres: int = 4
+) -> Optional[List[OpenAIFunction]]:
+    r"""Using semantic search to retrieve relevent tools
+    OpenAI / Open source Embedding models to embed the query and the schemas / docstring of `OpenAIFunction`
```
Comment (yiyiyi0817) [inline camel/functions/retrieve_tools_function.py:47]: each line should no more than 79 characters

File: camel/functions/retrieve_tools_function.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        sim_thres (int): 0.4 # The threshold above which we want to accept a tool
+
+    Returns:
+        retrieved_tools (List): List of retrieved (most relevant) tools.
+    """
+    # Check if k is greater than 0
+    assert k > 0, "k must be greater than 0"
+
+    # Check if tools is not empty
+    assert tools, "tools is empty"
```
Comment (yiyiyi0817) [inline camel/functions/retrieve_tools_function.py:63]: `if tools == []`, could we just return a empty list to let `retrieve_tools` more generalized? WDYT?

File: camel/functions/retrieve_tools_function.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    for func in tools:
+        text = func.func.__doc__ if func.func.__doc__ else ""
+        func_embedding = embed_text(text, model, tokenizer)
+        similarity = cosine_similarity(query_embedding, func_embedding)
+        similarities.append((similarity, func))
+
+    similarities.sort(key=lambda x: x[0], reverse=True)
+
+    if similarities[0][0] < sim_thres:
+        return None
```
Comment (yiyiyi0817) [inline camel/functions/retrieve_tools_function.py:88]: In the docstring, it said that return type is a List, so `return []` here? or revise "retrieved_tools (List): List of retrieved (most relevant) tools."

**Review by yiyiyi0817 — COMMENTED** [review]
Thanks @Hither1 ! Leave few comments.

File: camel/functions/retrieve_tools_function.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+def embed_text(text: str, model: AutoModel, tokenizer: AutoTokenizer):
+    r"""
+    Embed text using chosen model and tokenizer.
+    """
+    inputs = tokenizer(text, return_tensors="pt")
+    with torch.no_grad():
+        outputs = model(**inputs)
+
+    return outputs.last_hidden_state.mean(dim=1).squeeze().cpu().numpy()
```
Comment (zechengz) [inline camel/functions/retrieve_tools_function.py:31]: Some model may not use the mean pooling / aggregation. For example this model uses last token pooling https://huggingface.co/intfloat/e5-mistral-7b-instruct Also, one frequent text embedding is just to use the CLS token. IMO either you can use existing text embeddings https://github.com/camel-ai/camel/blob/master/camel/embeddings/sentence_transformers_embeddings.py or implement this in the embeddings folder with class name such as TransformersEncoder with different pooling / aggregation strategy

File: camel/functions/retrieve_tools_function.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+
+def cosine_similarity(vec1: np.ndarray, vec2: np.ndarray) -> float:
+    if np.linalg.norm(vec1) * np.linalg.norm(vec2) == 0:
+        return 0
+    else:
+        return np.dot(vec1, vec2) / (
```
Comment (zechengz) [inline camel/functions/retrieve_tools_function.py:38]: This may not be efficient, you can implement cosine similarity via l2-normalization https://stats.stackexchange.com/questions/146221/is-cosine-similarity-identical-to-l2-normalized-euclidean-distance

File: camel/functions/retrieve_tools_function.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+            np.linalg.norm(vec1) * np.linalg.norm(vec2)
+        )
+
+
+def retrieve_tools(
+    query: str, tools: List[OpenAIFunction], k: int = 3, sim_thres: int = 4
```
Comment (zechengz) [inline camel/functions/retrieve_tools_function.py:44]:
```suggestion
    query: str, tools: List[OpenAIFunction], k: int = 3, sim_thres: float = 0.4
```

**Review by zechengz — CHANGES_REQUESTED** [review]
Left some comments

**Review by onemquan — COMMENTED** [review]
please delete this empty line

File: camel/functions/__init__.py [hunk 19]
```diff
@@ -27,6 +27,7 @@
 from .weather_functions import WEATHER_FUNCS
 from .slack_functions import SLACK_FUNCS
 
+
```
Comment (onemquan) [inline camel/functions/__init__.py:30]: please delete this empty line

File: camel/functions/retrieve_tools_function.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+def cosine_similarity(vec1: List[float], vec2: List[float]) -> float:
+    r"""
+    Compute the cosine similarity between two sentence embedding vectors.
+
+    Args:
+        vec1 (List): the first sentence embedding vector.
+        vec2 (List): the second sentence embedding vector.
+    """
+    if np.linalg.norm(vec1) * np.linalg.norm(vec2) == 0:
+        return 0
```
Comment (onemquan) [inline camel/functions/retrieve_tools_function.py:33]: 0 => 0.0 to match the float type

File: camel/functions/retrieve_tools_function.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+        query (str): input question
+        tools (List): The tools to be searched.
+        k (int): how many answers the user wants to get.
+        sim_thres (int): 0.4 # The threshold above which we want to accept a tool
+
+    Returns:
+        retrieved_tools (List): List of retrieved (most relevant) tools.
+    """
+    # Check if k is greater than 0
+    assert k > 0, "k must be greater than 0"
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:59]: I would suggest using `if` statement for error handing rather than `assert`

File: camel/functions/retrieve_tools_function.py [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+    # Check not all descriptions are None
+    assert any(
+        func.func.__doc__ is not None for func in tools
+    ), "All docstrings are None"
+
+    # Check if sim_thres is between 0 and 1
+    assert sim_thres > 0 and sim_thres < 1, "sim_thres must be between 0 and 1"
+
+    # tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+    # model = AutoModel.from_pretrained("bert-base-uncased")
```
Comment (Wendong-Fan) [inline camel/functions/retrieve_tools_function.py:73]: please make the code clean if it's not needed

File: camel/toolkits/product_hunter.py
Comment (Wendong-Fan) [inline camel/toolkits/product_hunter.py:1]: seems the change for this file is not needed

**Review by Wendong-Fan — COMMENTED** [review]
left some comments~

File: camel/functions/retrieve_tools_function.py (hunk 21 above)
Comment (Wendong-Fan) [reply camel/functions/retrieve_tools_function.py:59]: refer: https://softwareengineering.stackexchange.com/questions/225956/python-assert-vs-if-return#:~:text=assert%20is%20intended%20for%20cases,strip%20it%20out%20in%20production.

File: camel/functions/retrieve_tools_function.py [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import List, Optional
+
+import numpy as np
+
+from camel.embeddings.sentence_transformers_embeddings import (
+    SentenceTransformerEncoder,
+)
```
[the remark this answers is not in the document]
Comment (Hither1) [reply camel/functions/retrieve_tools_function.py:20]: I don't really agree as some user may not have openai api key
